numbat/src/datetime.rs

```diff
@@ -1,82 +1,65 @@
             // 12 hour formats:
-            "%Y-%m-%d %I:%M:%S%.f %p",
+            "%Y-%m-%d %I:%M:%S %p", // TODO: add support for fractional seconds
             "%Y-%m-%d %I:%M %p",
-            "%Y/%m/%d %I:%M:%S%.f %p",
+            "%Y/%m/%d %I:%M:%S %p", // TODO: add support for fractional seconds
             "%Y/%m/%d %I:%M %p",
         ];
 
         for format in FORMATS {
             // Try to match the given format plus an additional UTC offset (%z)
-            if let Ok(dt) = chrono::DateTime::parse_from_str(input, &format!("{format} %z")) {
+            if let Ok(dt) = Zoned::strptime(&format!("{format} %z"), input) {
                 return Some(dt);
             }
 
             // Try to match the given format plus an additional timezone name (%Z).
-            // chrono does not support %Z, so we implement this ourselves. We were
+            // jiff does not support %Z, so we implement this ourselves. We were
             // warned by developers before us not to write timezone-related code on
             // our own, so we're probably going to regret this.
 
             // Get the last space-separated word in the input string, and try to parse it
             // as a timezone specifier, then try to match the rest of the string with the
             // given format.
             if let Some((rest, potential_timezone_name)) = input.rsplit_once(' ') {
-                if let Ok(tz) = potential_timezone_name.parse::<Tz>() {
-                    if let Ok(ndt) = chrono::NaiveDateTime::parse_from_str(rest, format) {
-                        if let LocalResult::Single(dt) = ndt.and_local_timezone(tz) {
-                            return Some(dt.fixed_offset());
-                        }
+                if let Ok(tz) = TimeZone::get(potential_timezone_name) {
+                    if let Ok(datetime) = DateTime::strptime(format, rest) {
+                        return Some(datetime.to_zoned(tz).unwrap()); // TODO
                     }
                 }
             }
 
             // Without timezone/offset
-            if let Ok(ndt) = chrono::NaiveDateTime::parse_from_str(input, format) {
-                if let LocalResult::Single(dt) = ndt.and_local_timezone(get_local_timezone_or_utc())
-                {
-                    return Some(dt.fixed_offset());
-                }
+            if let Ok(dt) = DateTime::strptime(format, input) {
+                return Some(dt.to_zoned(get_local_timezone_or_utc()).unwrap()); // TODO
             }
         }
 
         None
     }
 }
 
-pub fn to_rfc2822_save(dt: &DateTime<FixedOffset>) -> String {
-    if dt.year() < 0 || dt.year() > 9999 {
-        "<year out of range for displaying in RFC2822>".to_string()
+pub fn to_string(dt: &Zoned) -> String {
+    if dt.time_zone() == &TimeZone::UTC {
+        dt.strftime("%Y-%m-%d %H:%M:%S UTC").to_string()
     } else {
-        dt.to_rfc2822()
+        dt.strftime("%Y-%m-%d %H:%M:%S %Z (UTC %z)").to_string()
```

Comment:
This is something that was not (easily) possible before, I think. We can now display click times like CET or CEST instead of just the offsets. I chose to show both the clock time and the offset to the user for now. I personally don't know all of these abbreviations, so having the UTC offset in addition seems like a good idea.

Reply:
Yes, this is right. The abbreviation comes directly from your time zone database (i.e. `/usr/share/zoneinfo`).

If an abbreviation isn't available, then the offset is used. For example:

```
let zdt = date(2024, 7, 28).intz("Antarctica/Troll")?;
println!("{}", zdt.strftime("%F %Z"));
```

Gives this output:

```
$ cargo -q r
2024-07-28 +02
```

Reply:
> Antarctica/Troll

Ok, this is a really minor point.. but: I saw that in the docs as well, and initially interpreted it as some kind of non-existing timezone (troll as in: someone is trying to troll us). Might be just me, but could be worth using a different example (if one exists for the scenario in question).

> Yes, this is right. The abbreviation comes directly from your time zone database (i.e. `/usr/share/zoneinfo`).

Oh, good point. In that case, I would probably rather append `zoned_datetime.time_zone().iana_name()`, in case it exists. I don't want to show the offset twice.

Reply:
Ah I love `Antarctica/Troll` because it's notable for having a 2 hour DST transition! Hah. But yeah, I think I've heard someone else express confusion about that, so it's probably best to choose a different one. Thank you for the feedback.

> Oh, good point. In that case, I would probably rather append zoned_datetime.time_zone().iana_name(), in case it exists. I don't want to show the offset twice.

So if there was something like `%X` (specifier name to be determined) that expanded to the IANA time zone ID (and parsed it), what would it do if the IANA time zone ID wasn't available? I guess it would do the same as the abbreviation right? Probably just expand to the offset?

Reply:
> So if there was something like `%X` (specifier name to be determined) that expanded to the IANA time zone ID (and parsed it), what would it do if the IANA time zone ID wasn't available? I guess it would do the same as the abbreviation right? Probably just expand to the offset?

I think that is reasonable, because `%Z` does the same.

That said, my intention here in Numbat (and I guess in some other programs as well) is to show a `Zoned` datetime in a way that it is easy for humans to understand. And neither UTC offsets nor time zone IDs serve that purpose alone, in my opinion. Even for my local timezone, I would have to look up the precise DST rules to determine whether `2024-03-31 12:00:00 Europe/Berlin` would have a UTC offset of +01 or +02. So while this contains all information *in principle*, it's probably not ideal.

This is why I thought I'd maybe show both: timezone and UTC offset (and/or clock time). I don't care if it is verbose. Something like
```
2024-03-31 12:00:00 (UTC +0200, Europe/Berlin)
or even
2024-03-31 12:00:00 (CEST, UTC +0200, Europe/Berlin)
```

And for this use case, a fallback to the offset would not be ideal for `%Z`/`%X`. Because I would show the offset twice for cases where the timezone is not known (only the offset) or has no name.

I'm not sure if that makes sense. And if this is even the best way to present timestamps to a user. And I realize that the strftime/strptime grammar is (and should stay) somewhat limited, so it does not need to support each and every use case. I can also glue pieces together from various sources (such as `zoned.time_zone().iana_name()`).

Reply:
In terms of parsing, even with the time zone, the offset is actually required in order to losslessly format and parse the zoned datetime back into the same instant. So in terms of being verbose, I think you're heading in the right direction.

I'm not quite sure how to deal with "show offset and TZ id, but only show offset if there is no TZ id" in the context of the `strtime` APIs, since their expressivity is pretty limited. It would probably need a directive _just_ for that very specific formatting choice. The RFC 9557 interchange format deals with this by accepting redundancy. That is, `2024-07-28T14:16-04[-04]` is valid. I would be pretty tempted to ask you to consider redundancy to be okay in this case too. In general, it _should_ be rare for a TZ identifier to just not exist at all. It's _usually_ indicative of a system configuration problem _or_ a bug in Jiff (e.g., Jiff not _yet_ knowing how to detect the system default timezone on iOS or something).

Reply:
@BurntSushi I now wrote a little bit of custom logic to format datetimes for Numbat. Which is completely fine for me. `strftime` doesn't need to support each and every use case. I wanted it to look like the following. Note that there is no clock time (timezone abbreviation?) for Asia/Kathmandu:

![image](https://github.com/user-attachments/assets/a40d358e-1de6-4722-9b2a-c59bdcec9699)

The only thing which is totally hacky (and maybe not correct) is how I get the abbreviation and how I avoid printing the offset twice in case there is no abbreviation. I basically use `zoned.strftime("%Z")` to get the abbreviation as a string and then check if it starts with `+` or `-` (in which case I omit printing it). I guess it would be nice to have a way to get to that information without this hack, but I'm not sure if that's really a common use case.

https://github.com/sharkdp/numbat/blob/ad1543cf55d4021955f669899d6a0048cd5504b1/numbat/src/datetime.rs#L71-L95

Reply:
Yeah `Asia/Kathmandu` doesn't have any TZ abbreviations in tzdb:

```
[andrew@duff jiff]$ ./scripts/jiff-debug tzif /usr/share/zoneinfo/Asia/Kathmandu
TIME ZONE NAME
  /usr/share/zoneinfo/Asia/Kathmandu
TIME ZONE VERSION
  2
LOCAL TIME TYPES
  000:  offset=+05:41:16  designation=LMT        indicator=local/wall
  001:  offset=+05:30     designation=+0530      indicator=local/wall
  002:  offset=+05:45     designation=+0545      indicator=local/wall
TRANSITIONS
  0000:  -009999-01-02T01:59:59Z  unix=-377705023201  wall=-009999-01-02T07:41:15  unambiguous                      type=0  +05:41:16  LMT
  0001:  1919-12-31T18:18:44Z     unix=-1577943676    wall=1919-12-31T23:48:44     fold-until(1920-01-01T00:00:00)  type=1  +05:30     +0530
  0002:  1985-12-31T18:30:00Z     unix=504901800      wall=1986-01-01T00:00:00      gap-until(1986-01-01T00:15:00)  type=2  +05:45     +0545
POSIX TIME ZONE STRING
  <+0545>-5:45
```

> I basically use zoned.strftime("%Z") to get the abbreviation as a string and then check if it starts with + or - (in which case I omit printing it). I guess it would be nice to have a way to get to that information without this hack, but I'm not sure if that's really a common use case.

Yeah that's the [`TimeZone::to_offset`](https://docs.rs/jiff/latest/jiff/tz/struct.TimeZone.html#method.to_offset) API.

It just returns a `&str` though. It is quite literally passing through the abbreviation from tzdb itself. It isn't Jiff that's giving you the offset if there is no abbreviation. That comes from tzdb. So you'll still need to do the "check for +/- dance." But you at least don't need to go through `strftime`.

Your code looks correct though. I think what you're doing is sound and I can't think of any footguns here.

I am likely to add a `%V` specifier (from Java's [DateTimeFormatter](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/time/format/DateTimeFormatter.html#ISO_ZONED_DATE_TIME)) that will expand to an IANA tz identifier _or_ the UTC offset if the IANA tz identifier isn't available. But I don't think that will help you here to avoid the redundancy. I'll think on that a little more because I do like the format you settled on. It displays nicely.